GDScript: Display deprecated symbols in editor and language server

Meorge opened the pull request:
Related to https://github.com/godotengine/godot-proposals/issues/11079 !

This PR adds the ability for code completions from the language server to be marked as deprecated, according to the class documentation, so that they can be displayed as such in a code editor. It also implements this display for the built-in script editor:

In Godot's built-in script editor:
![CleanShot 2024-12-04 at 09 58 12@2x](https://github.com/user-attachments/assets/a74334b8-3c06-42a0-b03f-71f9f726d920)

In VS Code:
![CleanShot 2024-12-04 at 09 59 08@2x](https://github.com/user-attachments/assets/1d4f197f-48bd-43a6-b9a1-02f81fa567d0)

## Tested cases
### Methods
<details>

<summary>A user-defined method marked with `@deprecated`</summary>

```gdscript
## This method is deprecated.
## @deprecated
func do_something():
    print("Hello world")

func _ready():
	# do_something will show as deprecated
	do_som... 
```

</details>

<details>

<summary>A class built-in to Godot and marked as deprecated in the documentation</summary>

```gdscript
func _ready():
	var player = AnimationPlayer.new()
	# get_method_call_mode will show as deprecated
	player.get_method_c... 
```

</details>

### Classes

<details>

<summary>Main class of a file marked with `@deprecated`</summary>

```gdscript
class_name MyOldClass
extends Node

## This class is deprecated.
## @deprecated
```

```gdscript
# main.gd
extends Node

func _ready():
	# MyOldClass will show as deprecated
    var i = MyOl... 

```

</details>

<details>

<summary>Inner class marked with `@deprecated`</summary>

```gdscript
## This inner class is deprecated.
## @deprecated
class InnerDeprecatedClass:
	pass

func _ready():
	# InnerDeprecatedClass will show as deprecated
	var i = InnerDep...
```

</details>

<details>

<summary>A class built-in to Godot and marked as deprecated in the documentation</summary>

```gdscript
func _ready():
	# AnimatedTexture will show as deprecated
	var i = AnimatedTex...
```

</details>

<details>

<summary>A class built-in to Godot and marked as deprecated in the documentation, used as a type</summary>

```gdscript
func _ready():
	# AnimatedTexture will show as deprecated
	var i: AnimatedTex...
```

</details>

<details>

<summary>An autoload with a global variable name</summary>

```gdscript
# deprecated_autoload.gd
extends Node

## This autoload is registered with the name DeprecatedAutoload,
## and is marked as deprecated.
## @deprecated

func do_thing():
   print("a thing has been done")
```

```gdscript
# main.gd
func _ready():
    # DeprecatedAutoload will show as deprecated
    DeprecatedAu...
```

> [!WARNING]
> I've found this doesn't always work. It seems to involve something with the `.godot` folder preventing the doc data from being fully parsed, so I think it's outside the scope of this PR to try to fix.


</details>

### Enums and constants

<details>

<summary>
A user-defined constant, marked with `@deprecated`
</summary>

```gdscript
## This is a deprecated constant.
## @deprecated
const DEPRECATED_CONST = "hello"

func _ready():
	# DEPRECATED_CONST will show as deprecated
	print(DEPRECATED_CON...
```

</details>

<details>

<summary>A user-defined named enum, marked with `@deprecated`</summary>

```gdscript
## This is a deprecated enum.
## @deprecated
enum DeprecatedEnum { ONE, TWO, THREE }

func _ready():
	# DeprecatedEnum will show as deprecated
	print(Depre...
```

</details>

<details>

<summary>Individual values in a user-defined anonymous enum, marked with `@deprecated`</summary>

```gdscript
enum {
    ANONYMOUS_ONE,
    ## This one is deprecated
    ## @deprecated
    ANONYMOUS_TWO
}

func _ready():
	# ANONYMOUS_TWO will show as deprecated
	print(ANONYMOUS_T...
```

</details>

<details>

<summary>Individual values in a user-defined named enum, marked with `@deprecated`</summary>

```gdscript
enum NamedEnum {
	ONE,
	## This is a deprecated constant in an enum.
	## @deprecated
	TWO
}

func _ready():
	# NamedEnum.TWO will show as deprecated
	print(NamedEnum.TW... 
```

</details>

<details>

<summary>Individual values in an enum from a class built-in to Godot and marked as deprecated in the documentation</summary>

```gdscript
func _ready():
	# ANIMATION_PROCESS_PHYSICS will show as deprecated
	var i = AnimationPlayer.ANIMATION_PROCESS_PH...

	# Will also show as deprecated
	i = AnimationPlayer.AnimationProcessCallback.ANIMATION_PROCESS_PH...
```

</details>

<details>

<summary>Constants in `@GlobalScope`</summary>

```gdscript
func _ready():
	# PROPERTY_HINT_NODE_PATH_TO_EDITED_NODE will show as deprecated
	print(PROPERTY_HINT_NODE_PATH_T...
```

</details>

### Properties

<details>

<summary>A user-defined property marked with `@deprecated`</summary>

```gdscript
## This is a deprecated property.
## @deprecated
var deprecated_property: float = 1.0

## This is a deprecated property with a getter and setter.
## @deprecated
var deprecated_property_2: float = 2.0:
    get:
        return sqrt(5)
    set(value):
        deprecated_property_2 = value

func _ready():
	# deprecated_property and deprecated_property_2 will show as deprecated
	print(deprecated_prop...
```

</details>

<details>

<summary>A class built-in to Godot and marked as deprecated in the documentation</summary>

```gdscript
func _ready():
	# auto_translate will show as deprecated
    var i = Control.new().auto_tr...
```

</details>

### Signals

<details>

<summary>A user-defined signal marked with `@deprecated`</summary>

```gdscript
## This is a deprecated signal.
## @deprecated
signal deprecated_signal

func _ready():
	# deprecated_signal will show as deprecated
	depreca...
```

</details>

<details>

<summary>A class built-in to Godot and marked as deprecated in the documentation</summary>

```gdscript
func _ready():
	# setup_local_to_scene_requested will show as deprecated
	Resource.new().setup_local_t...
```

</details>

## To-Do
### Language server
- [x] Fix behavior with autoloads (determine how to retrieve their class's `DocData`)
- [x] General clean-up of code structure - it's messy right now! (Maybe add some static methods for getting is-deprecated hashmaps?)
   - I've tried to standardize the variable names used so I think it's better now, but would still appreciate suggestions if you have them!
### Godot Script editor
- [x] Ensure the strikethrough scales correctly as the editor scale is changed

Meorge committed:
GDScript: Display deprecated symbols in editor and language server

Meorge commented:
It seems like just about everything is working, **except** for global variable names for autoloads. As stated above:

> [!WARNING]
> Autoload global variable names will *not* display as deprecated.
> ```gdscript
> # deprecated_autoload.gd
> extends Node
> 
> ## This autoload is registered with the name DeprecatedAutoload,
> ## and is marked as deprecated.
> ## @deprecated
> 
> func do_thing():
>    print("a thing has been done")
> ```
> 
> ```gdscript
> # main.gd
> func _ready():
>     # DeprecatedAutoload will *not* show as deprecated
>     DeprecatedAu...
> ```

I've spent a fair amount of time this evening trying to figure out how to get to the `DocData::ClassDoc` for an autoload, and based on the existing code it looks like I'd need some kind of function that maps script paths to `ClassNode`s (or `DocData::ClassDoc`s).

**What are others' thoughts about not displaying global variable names for autoloads as deprecated even if they're marked as deprecated in their GDScript file's documentation comments?**

Meorge commented:
> From what I see autoload doc data is also registered in EditorHelp::get_doc_data().class_list under the name of the autoload, but the update behavior seems very flaky so I'm not sure whether we can rely on it.

I'm finding this to be shaky as well. I have a test project that I'm using, and can do the following:

1. I delete the project's `.godot` folder, then open the project. If I go straight to "Search Help" and start typing the name of the global variable the autoload is registered under, it appears in the list.
2. I make *a* change to a script file in the project and save it, then close the project.
3. I re-open the project and go to "Search Help" again, and start typing the name of the global variable the autoload is registered under. It doesn't appear anymore.

Indeed, using the following snippet:
```cpp
for (const KeyValue<String, DocData::ClassDoc> &E2 : EditorHelp::get_doc_data()->class_list) {
    print_line(vformat("Class %s found", E2.key));
}
```
I do get the following output:
```
...
Class PackedColorArray found
Class PackedVector4Array found
Class @GlobalScope found
Class @GDScript found
Class DeprecatedAutoload found
Class DeprecatedClass found
Class MyScene.InnerDeprecatedClass found
Class MyScene found
```
(where `DeprecatedAutoload` is the name of the autoload's global variable, *not* its `class_name`).

And indeed, if I delete `.godot` then open the project, I *can* get the `DeprecatedAutoload` to display as deprecated:

![CleanShot 2024-12-07 at 11 46 48](https://github.com/user-attachments/assets/19798fde-591f-496f-aedf-ef3e310471ee)

So it seems to me like something being cached in `.godot` is causing the doc generation to halt prematurely, such that autoloads aren't being added. Last night as I was working on this, I noticed my other custom classes also weren't printing out. I'll look into getting an issue filed for this, if one doesn't already exist!

---

Regarding performance, I agree that's definitely a big concern, and one I've been conscious of as I've been working through putting this feature together. Across the file, there are a few ways I've gotten deprecated data from the docs:

When I have the `Node` for a symbol, I can grab the data from it directly via something like `member.m_class->doc_data.is_deprecated`. As far as I can tell, this would have very little overhead.

For checking if a class is deprecated, I have to get `EditorHelp::get_doc_data()->class_list` and access it by key. This seems like it might have a bit more overhead.

So far, I'm most nervous about when I want to check whether signals/properties/methods on some class are deprecated. My process so far has been to:
1. Get the `ClassDoc` for the given class.
2. Create an empty `HashMap<StringName, bool>` to store names and whether or not they are deprecated.
3. Iterate over the relevant vector within the `ClassDoc`, and for each item, add an entry to the `HashMap` where the key is the item's name and the value is whether or not the item is deprecated.
4. When creating the vector of `CodeCompletionOption`s, check if each one exists in the `HashMap`, and if so, set its `deprecated` value to the corresponding value in the `HashMap`.

In practice, that looks like this:
```cpp
// Up at the top of the method, so it can be reused:
const HashMap<String, DocData::ClassDoc> class_doc_map = EditorHelp::get_doc_data()->class_list;

// ...

// Here we're going to create the list of constants to provide as autocomplete suggestions.
HashMap<StringName, Variant> constants;
scr->get_constants(&constants);

// Create a HashMap mapping constants' names to their deprecated status.
HashMap<StringName, bool> const_is_deprecated_map;
for (const DocData::ConstantDoc &constant_doc : class_doc_map.get(base_type.class_type->get_global_name()).constants) {
    const_is_deprecated_map.insert(constant_doc.name, constant_doc.is_deprecated);
}

for (const KeyValue<StringName, Variant> &E : constants) {
    int location = p_recursion_depth + _get_constant_location(scr, E.key);
    ScriptLanguage::CodeCompletionOption option(E.key.operator String(), ScriptLanguage::CODE_COMPLETION_KIND_CONSTANT, location);

    // Add the constant's deprecated status if we have it.
    if (const_is_deprecated_map.has(E.key.operator String())) {
        option.deprecated = const_is_deprecated_map.get(E.key.operator String());
    }
    r_result.insert(option.display, option);
}
```

Without changing how the doc data is structured, I'm not sure how this could be made much more efficient. These hashmaps could perhaps be computed earlier on, alongside the rest of the doc data, and then cached? Although then that introduces multiple sources of truth for deprecated statuses, which isn't ideal.

HolonProduction commented:
> For checking if a class is deprecated, I have to get `EditorHelp::get_doc_data()->class_list` and access it by key. This seems like it might have a bit more overhead.

I wouldn't worry about hashmap lookups and such, that should be fine. Just triggering parsing and analyzing for a script can get very expensive.

HolonProduction commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1280,6 +1398,9 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 				ClassDB::get_enum_list(type, &enums);
 				for (const StringName &E : enums) {
 					int location = p_recursion_depth + _get_enum_location(type, E);
+					// Enum names for a native class, including `@GlobalScope`.
```
Testing on recent dev I can't get any enum names from `@GlobalScope` to show at all so I'm not sure whether this is correct, and whether they should be handled here if it were to work.

HolonProduction replied:
Thinking about it this would probably be the right place. Maybe we should add a TODO comment somewere or I'll create an issue for it.

HolonProduction commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -2722,19 +2995,27 @@ static bool _guess_method_return_type_from_base(GDScriptParser::CompletionContex
 		if (p_context.current_class && p_context.current_class->has_member(current_enum) && p_context.current_class->get_member(current_enum).type == GDScriptParser::ClassNode::Member::ENUM) {
 			const GDScriptParser::EnumNode *_enum = p_context.current_class->get_member(current_enum).m_enum;
 			for (int i = 0; i < _enum->values.size(); i++) {
+				// TODO: Determine what case this covers.
```
Enum members from the current GDScript, specific to a certain context. So when a type parameter of a function requires an enum this method gets called an will search possible enum values for this specifc enum. (Also happens with assignments IIRC)

HolonProduction replied:
Ok so I looked again. For function calls it only gets called for functions of native types (e.g. `Control.set_anchor`). Might be something to extend to CLASS types though, since the function would be able to handle it from what I see. For assignments it could get called for class types, but in practice never is, since the type attached to gdscript variables doesn't have the enumeration name set (might be a bug don't know). So yeah for now it is only ever called for native functions and properties.

HolonProduction commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -2722,19 +2995,27 @@ static bool _guess_method_return_type_from_base(GDScriptParser::CompletionContex
+			ERR_FAIL_COND_MSG(!class_list.has("@GlobalScope"), vformat("@GlobalScope class couldn't be found in doc data by the language server"));
 			for (int i = 0; i < CoreConstants::get_global_constant_count(); i++) {
 				if (CoreConstants::get_global_constant_enum(i) == current_enum) {
+					// TODO: Determine what case this covers.
```
Global constant enum values (e.g. `SIDE_LEFT`) for a specifc that has a global enum type (e.g. `Side`).

Meorge replied:
Similar to above, this doesn't seem to be triggering for me (no output from the `print_line` statement I added to this branch's `option` either):

![CleanShot 2024-12-12 at 10 16 21@2x](https://github.com/user-attachments/assets/372d38b2-284d-462c-94d1-121e2c3721da)

HolonProduction replied:
Same as above only works for native methods and properties. I guess I'll go through this again and create a lot of issues xD

HolonProduction commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -2749,10 +3030,23 @@ static void _find_enumeration_candidates(GDScriptParser::CompletionContext &p_co
 		for (const StringName &E : enum_constants) {
 			String candidate = class_name + "." + E;
 			int location = _get_enum_constant_location(class_name, E);
+			// TODO: Determine what case this covers.
```
Native enum values for a context which has a native enum type.

Meorge replied:
Is this a case where it would cannot currently be triggered, but would be if/when this method is called on Script types as well as native methods? And/or do you have an example of GDScript code that would trigger this completion branch?

HolonProduction replied:
This can be triggered given the following script:
```gdscript
extends Control

func _init():
    set_anchors_and_offsets_preset(...)
```

Meorge replied:
Thanks! Example has been added 😄

HolonProduction commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -916,6 +916,13 @@ static const int PlainText = 1;
 static const int Snippet = 2;
 }; // namespace InsertTextFormat
 
+namespace CompletionItemTag {
+/**
+ * Render a completion as obsolete, usually using a strike-out.
+ */
+static const int Deprecated = 1;
+}; // namespace CompletionItemTag
+
```
We don't have client capability tracking, tags are only available from LSP 3.15.0. IMO we should stick with `deprecated`.

Meorge replied:
According to the [official specification](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionItem), it sounds like `CompletionItem.deprecated` is itself deprecated:
```typescript
export interface CompletionItem {
	...
	/**
	 * Indicates if this item is deprecated.
	 *
	 * @deprecated Use `tags` instead if supported.
	 */
	deprecated?: boolean;
	...
}
```
I thought I checked and found that VS Code, at least, didn't support `CompletionItem.deprecated`, but I can take another look. 

It looks like we may also be able to check the client's capabilities via [`CompletionClientCapabilities`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionClientCapabilities), including whether or not they support either `CompletionItem.deprecated` or `CompletionItemTag.Deprecated`.

HolonProduction replied:
> It looks like we may also be able to check the client's capabilities via [`CompletionClientCapabilities`](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#completionClientCapabilities), including whether or not they support either `CompletionItem.deprecated` or `CompletionItemTag.Deprecated`.

I know, that's how it's supposed to work. But our server implementation doesn't implement capability tracking. We just ignore the capabilities that a client sends. I was hoping `deprecated` would work even if it is deprecated in the protocol. When we add tags without proper capability tracking we risk breaking other client implementations that aren't up to date.

I originally didn't want to implement capability tracking before limiting clients to 1 client per server (see https://github.com/godotengine/godot-proposals/issues/11056). But if this becomes a blocker it should be possible to just save them per client for now (at the risk of breaking some assumptions when multiple clients are connected).

Meorge replied:
Great news on this - looks like VS Code does still support the `deprecated` field, without using the tags, unless there's some building/caching weirdness going on. I commented out all of the instances of `tags` and replaced them with the `deprecated` field, and it seems to still be doing the strikethrough.

For now, I'll remove the tags entirely and just use the `deprecated` field :)

HolonProduction commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -931,6 +938,11 @@ struct CompletionItem {
+	/**
+	 * Tags for this completion item.
+	 */
+	Array tags;
```
See above

Meorge reviewed (commented):
Some of the behaviors I tried to verify with tests but wasn't able to. Specifically, it seemed like `_find_enumeration_candidates` was never being triggered. They may well be bugs we should file separate issues for.

HolonProduction commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1025,18 +1038,33 @@ static void _list_available_types(bool p_inherit_only, GDScriptParser::Completio
 					} break;
 					case GDScriptParser::ClassNode::Member::CONSTANT: {
 						if (member.constant->get_datatype().is_meta_type && p_context.current_class->outer != nullptr) {
+							// TODO: Describe this case and its examples.
```
Yeah forgot, it's broken on `master`. https://github.com/godotengine/godot/pull/94996 would fix it but I can't get any reviews it seems :(

Funnily enough the PR fixes this in a different code path. Might need to investigate this one again.

HolonProduction commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1105,7 +1144,12 @@ static void _find_identifiers_in_class(const GDScriptParser::ClassNode *p_class,
 						if (p_types_only || p_only_functions || outer || (p_static && !member.variable->is_static)) {
 							continue;
 						}
+						// Property of a class, when used in an expression.
```
In general some of those methods are very versatile in what they can handle. So documenting them in this way which is dependent on the call context seems like the wrong direction. We should document what those methods do and which parameters are fine. With this documentation it becomes clear were they can be used.

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1186,14 +1260,21 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 	while (!base_type.has_no_type()) {
+		// This HashMap is declared up here because it works across both
+		// the ENUM and BUILTIN cases.
+		HashMap<String, bool> enum_val_is_deprecated_map;
```
For these I'd instead suggest using `HashSet<String>` and inserting if deprecated, and then using `has`, that avoids storing excessive entries

Meorge replied:
Thanks for the suggestion, that makes a lot of sense! 😄

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1209,6 +1290,12 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 							List<PropertyInfo> members;
 							scr->get_script_property_list(&members);
+
+							HashMap<StringName, bool> prop_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1217,23 +1304,50 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 							List<MethodInfo> signals;
 							scr->get_script_signal_list(&signals);
+
+							HashMap<StringName, bool> signal_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1217,23 +1304,50 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 						HashMap<StringName, Variant> constants;
 						scr->get_constants(&constants);
+
+						HashMap<StringName, bool> const_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1290,16 +1409,41 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 				if (!p_only_functions) {
 					List<String> constants;
+
+					HashMap<String, bool> const_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1290,16 +1409,41 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 						List<PropertyInfo> pinfo;
 						ClassDB::get_property_list(type, &pinfo);
+
+						HashMap<String, bool> prop_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1308,15 +1452,39 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 						List<MethodInfo> signals;
 						ClassDB::get_signal_list(type, &signals);
+
+						HashMap<String, bool> signal_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1326,6 +1494,17 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 				ClassDB::get_method_list(type, &methods, false, true);
+
+				// Assemble a hashmap of method names and whether or not they're deprecated.
+				HashMap<String, bool> method_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1376,21 +1584,30 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 
 				if (!p_only_functions) {
 					List<PropertyInfo> members;
+					HashMap<String, bool> member_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1402,12 +1619,27 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 				List<MethodInfo> methods;
 				tmp.get_method_list(&methods);
+
+				HashMap<String, bool> method_is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1503,7 +1742,17 @@ static void _find_identifiers(const GDScriptParser::CompletionContext &p_context
+	// Build a hashmap of constants from @GlobalScope,
+	// and whether or not they are deprecated.
+	ERR_FAIL_COND_MSG(!class_list.has("@GlobalScope"), vformat("@GlobalScope class couldn't be found in doc data by the language server"));
+	HashMap<String, bool> const_is_deprecated;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -2749,10 +3034,24 @@ static void _find_enumeration_candidates(GDScriptParser::CompletionContext &p_co
+
+		// Build hashmap of enum constants and whether or not they're deprecated
+		ERR_FAIL_COND_MSG(!class_list.has(class_name), vformat("%s class couldn't be found in doc data by the language server", class_name));
+		HashMap<String, bool> const_is_deprecated;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3237,20 +3538,44 @@ ::Error GDScriptLanguage::complete_code(const String &p_code, const String &p_pa
+				// Create hashmap of constants within this Variant, and whether
+				// or not they are deprecated.
+				// ERR_FAIL_COND_MSG(!class_list.has(variant_name), vformat("Variant \"%s\" couldn't be found in doc data by the language server", variant_name));
+				HashMap<String, bool> is_deprecated_map;
```
Same

AThousandShips commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -3237,20 +3538,44 @@ ::Error GDScriptLanguage::complete_code(const String &p_code, const String &p_pa
+				// Create hashmap of methods within this Variant, and whether
+				// or not they are deprecated.
+				// ERR_FAIL_COND_MSG(!class_list.has(variant_name), vformat("Variant \"%s\" couldn't be found in doc data by the language server", variant_name));
+				HashMap<String, bool> is_deprecated_map;
```
Same

Meorge commented:
There was an issue where I was attempting to access editor theming, and causing some of the CI to fail, but I *think* I found a fix for that. If all the checks don't pass, I'll try to continue working through them on another branch so that they put less strain on the main repo's CI resources 😅

dalexeev commented:
Just a note that after #91060 local variables and constants can also be deprecated/experimental. Not sure how useful this is in practice and whether LSP supports deprecating local identifiers, but it would be nice to add it.

Meorge commented:
For me, *practically* speaking, it doesn't make sense for a local variable to be deprecated - if so then I feel like it's time for a refactor of your code 😅 But, if documentation comments are parsed for local identifiers and that includes deprecated status, I also agree that for consistency's sake it should be added.

And added it is! I may not push it to the branch that this PR is tracking immediately, as I'd like to get the CI issues resolved first.

![CleanShot 2024-12-20 at 11 03 20@2x](https://github.com/user-attachments/assets/fd28e6e9-2292-475a-a185-096df49f9082)
![CleanShot 2024-12-20 at 11 03 41@2x](https://github.com/user-attachments/assets/b2a0d0d2-0199-452f-aaff-4a6c46088362)

dalexeev commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1209,6 +1294,16 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 							scr->get_script_property_list(&members);
+
+							HashSet<StringName> deprecated_properties;
+							if (base_type.class_type && class_doc_map.has(base_type.class_type->get_global_name())) {
```
I think you shouldn't use `get_global_name()` to determine the doc class name. Because for unnamed classes, inner classes and autoload singletons the name may differ from the global name.

After #91060 there is a new method `Script::get_doc_class_name()`, but there is no counterpart for `GDScriptParser::ClassNode`. I reused `GDScriptDocGen::doctype_from_gdtype()` (which uses `GDScriptDocGen::_get_class_name()`).

Meorge replied:
Thanks! I'll look more into this once I have the compatibility method stuff working and CI tests passing.

Meorge commented:
I'm currently working on getting the CI checks to pass. By adding the `deprecated` argument to `CodeEdit::add_code_completion_option`, I broke compatibility for that method and am now trying to get a compatibility method for it to work correctly. To avoid using Godot's CI resources but still run the checks, I'm continuing it on the branch at https://github.com/Meorge/godot/tree/language-server-tomfoolery-dev. It seems to be failing with the "Hash changed" error.

In `code_edit.compat.inc`:
```cpp
void CodeEdit::_add_code_completion_option_compat_100019(CodeCompletionKind p_type, const String &p_display_text, const String &p_insert_text, const Color &p_text_color, const Ref<Resource> &p_icon, const Variant &p_value, int p_location) {
	add_code_completion_option(p_type, p_display_text, p_insert_text, p_text_color, p_icon, p_value, p_location, false);
}

void CodeEdit::_bind_compatibility_methods() {
	// ...
	ClassDB::bind_compatibility_method(D_METHOD("add_code_completion_option", "type", "display_text", "insert_text", "text_color", "icon", "value", "location"), &CodeEdit::_add_code_completion_option_compat_100019, DEFVAL(Color(1, 1, 1)), DEFVAL(Ref<Resource>()), DEFVAL(Variant::NIL), DEFVAL(LOCATION_OTHER), DEFVAL(false));
}
```

In `code_edit.h`:
```cpp
#ifndef DISABLE_DEPRECATED
	// ...
	void _add_code_completion_option_compat_100019(CodeCompletionKind p_type, const String &p_display_text, const String &p_insert_text, const Color &p_text_color = Color(1, 1, 1), const Ref<Resource> &p_icon = Ref<Resource>(), const Variant &p_value = Variant::NIL, int p_location = LOCATION_OTHER);
	static void _bind_compatibility_methods();
#endif
```

And, in `4.3-stable.expected`:
```
GH-100019
--------
Validate extension JSON: Error: Field 'classes/CodeEdit/methods/add_code_completion_option/arguments': size changed value in new API, from 7 to 8.

New argument `deprecated` added to `add_code_completion_option`. Compatibility method registered.
```

However I'm still getting the error:
```
Compatibility to 4.3-stable is broken in the following ways:
Validate extension JSON: Error: Hash changed for 'classes/CodeEdit/methods/add_code_completion_option', from EB1A746E to 611C3D20. This means that the function has changed and no compatibility function was provided.
```

I feel like I've followed all of the steps at https://docs.godotengine.org/en/stable/contributing/development/handling_compatibility_breakages.html, so I'm feeling a bit out of ideas at the moment... 😅

Meorge commented:
@HolonProduction It should be ready for re-review! I've implemented dalexeev's suggestion for `get_doc_class_name()` and rebased it to a more recent commit 😄

Calinou commented on doc/classes/CodeEdit.xml:
```diff
@@ -50,9 +50,11 @@
 			<description>
 				Submits an item to the queue of potential candidates for the autocomplete menu. Call [method update_code_completion_options] to update the list.
 				[param location] indicates location of the option relative to the location of the code completion query. See [enum CodeEdit.CodeCompletionLocation] for how to set this value.
+				If [param deprecated] is `true`, then the item will be indicated to the user as being unwise to use (such as with a strikethrough effect).
```
```suggestion
				If [param deprecated] is [code]true[/code], then the item will be indicated to the user as being unwise to use (such as with a strikethrough effect).
```

Calinou reviewed (commented):
Tested locally, it works as expected with all symbol types (I've tested both enums and enum values too). Code looks good to me.

One issue I noticed is that the strikethrough is drawn too low and almost looks like an underline:

![image](https://github.com/user-attachments/assets/37442b5e-55c5-4cac-b526-c6f5f56a7e4c)

See RichTextLabel's strikethrough drawing guide for a reference:

https://github.com/godotengine/godot/blob/7598b08ec2ccf6c5bbc2c2020f583fc625e3238b/scene/gui/rich_text_label.cpp#L1159-L1165

Additionally, you can get the font's underline width from the font metadata to use a line width (this is already done in RichTextLabel for both strikethrough and underline):

https://github.com/godotengine/godot/blob/7598b08ec2ccf6c5bbc2c2020f583fc625e3238b/scene/gui/rich_text_label.cpp#L939-L943

Meorge commented:
Thanks for testing it! I've been working at trying to improve the strikethrough for the editor, and currently have this:

![CleanShot 2025-03-28 at 15 30 46@2x](https://github.com/user-attachments/assets/db354155-120e-40c6-a367-b5a4ed26f097)
![CleanShot 2025-03-28 at 15 32 26@2x](https://github.com/user-attachments/assets/c0613cc4-eca3-44ec-8bb5-12a7f9bbab58)

And, with `example()` to match your image (old, then new):

![CleanShot 2025-03-28 at 15 30 30@2x](https://github.com/user-attachments/assets/e4e17a62-53de-440c-86d4-379c91a64118)
![CleanShot 2025-03-28 at 15 27 32@2x](https://github.com/user-attachments/assets/b13fec89-9f87-44b8-a82e-c74063668919)

The Y position of the strikethrough is calculated as
```cpp
double l_ascent = TS->shaped_text_get_ascent(rid);
float st_width = MAX(1.0, TS->shaped_text_get_underline_thickness(rid) * get_theme_default_base_scale());

double y_pos = title_pos.y + l_ascent / 2 + st_width / 2
```
I'm worried the line may now be up a little too high, but the logic here feels sound. The `title_pos.y` should be the position of the top edge of the text, and `l_ascent / 2` should take it down to halfway through the baseline, where we presumably want it. Adding half the strikethrough width should move it down even further.

I'll push this updated version so you can take a look at it, if you're able. The `example()` one drew differently for me than for you, so it may just be a system difference?

Calinou reviewed (approved):
Looks great now :slightly_smiling_face: 

![image](https://github.com/user-attachments/assets/ec5608f2-16df-4b6e-bd17-335d8e4cd0c1)

And at 200% editor scale too:

![image](https://github.com/user-attachments/assets/ea60723d-dce1-4bcf-8243-2fa0a8cbbb64)

Meorge commented:
Thanks! I just realized, there's a stray `[no ci]` in one of the commits... I'll get that removed and the branch re-pushed 😅

AThousandShips commented on misc/extension_api_validation/4.4-stable.expected:
```diff
@@ -24,3 +24,11 @@ Validate extension JSON: Error: Field 'classes/OpenXRAPIExtension/methods/regist
+
+
+GH-100019
+--------
```
```suggestion
---------
```

HolonProduction reviewed (commented):
First of all I tested this a bit and it seems to work correctly :+1:  Skimming over the GDScript code, it seems fine on the small scale.

That being said, I don't like the approach this PR takes from an architectural standpoint.

This code adds too much coupling between GDScript and the editor. There exists a well defined API for how the GDScript module interacts with the editor in  `ScriptLanguage` other than that they should probably not know about each other, this includes things like `EditorHelp`.

Also the GDScript module has a well defined job: handling GDScript related things. Interacting with the documentation cache most certainly isn't GDScript related so this breaks cohesion IMO. I don't see why we have to iterate and join the documentation in GDScript specific code. There were no usages of `EditorHelp` in this file before, even the documentation tooltip feature managed to keep those concerns separated and only resolves documentation for locals in the GDScript module.

In my opinion we should attach symbol information to the completion options (the ones in the script language interface, not the ones in `CodeEdit`), the editor can use this symbol information to look for the correct documentation.

That's easier said than done, since there is no unified symbol model that would support locals at the moment. So please don't try to implement this! I'm just trying to say that the script language interface might need some refactoring to facilitate the addition of more IDE like features.

I know that Godot has this ["one problem one solution"](https://docs.godotengine.org/en/stable/contributing/development/best_practices_for_engine_contributors.html#to-each-problem-its-own-solution) policy, and in this case I disagree with it, a lot of IDE like features share the need to work with certain symbols in one way or the other, encapsulating this in a flexible model and using it to implement such features seems like the best approach to me. That being said I acknowledge that this policy has brought Godot to were it is today, so if the team wants to move forward with this feature like it is implemented here, that's fine with me. The specific code in the functions looks mostly good and _should_ be fine to merge :)

I'll neither approve nor reject this, I'll just leave my opinion here, someone more familiar with the project should make this decision.

Meorge commented on modules/gdscript/language_server/godot_lsp.h:
```diff
@@ -930,6 +930,11 @@ struct CompletionItem {
 	 */
 	int kind = 0;
 
+	/**
+	 * Indicates if this item is deprecated.
+	 */
+	bool deprecated = false;
+
```
Good question, I'm not sure why myself. I see it is indeed after `documentation` and before `preselect`, so I'll move it there.

Meorge commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1004,22 +1005,36 @@ static void _list_available_types(bool p_inherit_only, GDScriptParser::Completio
+			// Native class, when used as a type.
+			// 		var my_var: No...  # <-- will suggest `Node2D`
 			ScriptLanguage::CodeCompletionOption option(E, ScriptLanguage::CODE_COMPLETION_KIND_CLASS);
+			if (class_doc_map.has(E)) {
+				option.deprecated = class_doc_map.get(E).is_deprecated;
+			}
```
This was mostly just done to keep the line length more reasonable and readable. In this case here, it looks like I also gave it a default empty HashMap so that if the class list wasn't created yet, it wouldn't cause an error.

The way that such error handling is done probably varies across this PR, as I was learning how to access it on the spot, so it may be worth me deciding on a single approach, then going through and making sure it's applied everywhere 😅

Meorge commented:
Thanks for all of the comments and suggestions! I have some other things I need to work on over the next few days, but I'm hoping to spend more time on this PR soon.

I definitely get what you mean about the coupling of GDScript and `EditorHelp`. My PR for deprecated symbol warnings, #104955 , is even worse in this regard IMO, as it links the analyzer to `EditorHelp`. Since these features uniquely tie documentation info to code parsing/analysis itself, I feel like in the long run it would make sense to decouple the documentation info from the editor and allow both the editor and analyzer to reference it. If that idea even makes sense to the experienced GDScript people, though, I recognize it would be quite a task to implement 😅

Meorge commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1030,18 +1045,31 @@ static void _list_available_types(bool p_inherit_only, GDScriptParser::Completio
+							// const MyType = preload("res://my_type.gd)
+							// var m: MyT... # <-- will suggest MyType
 							ScriptLanguage::CodeCompletionOption option(member.constant->identifier->name, ScriptLanguage::CODE_COMPLETION_KIND_CLASS, ScriptLanguage::LOCATION_LOCAL);
+							option.deprecated = member.constant->doc_data.is_deprecated;
```
Ah yes, I definitely see what you're saying there. I'll look into how to find the actual script class from the information available to the code here.

Meorge commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1366,9 +1565,33 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
+					}
+				} else {
+					if (!class_doc_map.has(type)) {
+						print_line(vformat(R"(Class "%s" couldn't be found in doc data by the language server.)", type));
```
As I recall, `class_doc_map` just has the class names themselves, not the enum names. `type` from above is defined as `base_type.native_type.get_slicec('.', 0)` so that if statement is just checking if the first component of the type string (the class name) is in `class_doc_map`.

Mickeon replied:
Add this as more fuel to the fire @dalexeev https://github.com/godotengine/godot/pull/98233

Meorge commented on modules/gdscript/gdscript_editor.cpp:
```diff
@@ -1301,16 +1433,43 @@ static void _find_identifiers_in_base(const GDScriptCompletionIdentifier &p_base
 					List<String> constants;
+
+					HashSet<String> deprecated_consts;
+					if (!class_doc_map.has(type)) {
+						print_line(vformat(R"(Class "%s" couldn't be found in doc data by the language server.)", type));
```
I've removed the "by the language server" in all of the messages. As for the printing function/macro to use, I wasn't sure if it made sense to have them abort the current method. Since this is an optional feature, I was thinking `print_error()` might make more sense - that way, AFAIK, it makes it clear that something went wrong to the user, but it also doesn't break anything else.

Meorge replied:
Looks like using `ERR_FAIL_COND_MSG` was causing some tests to fail, so I reverted them to `print_error()` as well

Meorge commented:
Looking through #101370 some more, I feel like I understand a bit better how these two PRs would fit together. I don't think the other PR would change much about how this PR would function.

The general flow, as I understand it, is:
- The code structure is analyzed for potential completion candidates. This is where **this PR** adds the information about whether or not something is deprecated.
- If the user is typing arguments for a function, then #101370 uses the information from the previous step to look up and display the relevant documentation.

Because this PR performs its actions before #101370, merging the latter first wouldn't change how this one would work. However, the deprecation status data added to completion options by this PR might be able to be incorporated into #101370 to display whether or not the function a user is currently trying to use is deprecated.

AdriaandeJongh reviewed (approved):
I haven't looked at the code, but from a usability perspective this looks good. I was not able to check all the possible usages of deprecation supported by this PR, but the few I did try worked well. I think the place where most users will come across this is deprecated engine functions, and those show up correctly as strikethrough in autocomplete now.

Meorge commented:
Another PR of mine, #104955, adds some dedicated helper methods that can look up deprecation info for given symbols/names. Once it is merged, I can adapt this PR's code to use those methods. (Alternatively, if it feels like this PR is closer to merging than the other one, I can copy the helper methods over to this PR.)
